File: P1.py
# -*- coding: utf-8 -*-
"""
Created on Mon Apr 19 10:19:06 2021

@author: jgdch
"""

#Time Complexity:O(N), N is the number of ndoes in the tree
#Space Complexity:O(N)
#Passed all LC Test testcases

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
        self.index=0
        self.hmap={}
   
    def buildTree(self, inorder: List[int], postorder: List[int]) -> TreeNode:
        self.index=len(postorder)-1
        for i in range(len(inorder)):
            self.hmap[inorder[i]]=i
        
        return self.helper(postorder,0,len(postorder)-1)
    def helper(self,postorder,start,end):
        if(start>end):
            return None
        root=TreeNode(postorder[self.index])
        self.index-=1
        ind=self.hmap[root.val]
        root.right=self.helper(postorder,ind+1,end)
        root.left=self.helper(postorder,start,ind-1)
        return root
        
        

File: p1.java
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :No

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    // Hashmap to keep track of values and index in inorder array
    Map<Integer, Integer> map;
    // pointer for postorder array
    int idx;

    public TreeNode buildTree(int[] inorder, int[] postorder) {
        // postorder pointer will start from n-1
        idx = postorder.length - 1;
        map = new HashMap<>();

        // Putting all elements and it index into hashmap from inorder
        for (int i = 0; i < inorder.length; i++) {
            map.put(inorder[i], i);
        }

        // Calling recursive function which will return root node
        return helper(postorder, 0, inorder.length - 1);
    }

    public TreeNode helper(int[] postorder, int start, int end) {
        // base condition
        if (start > end)
            return null;

        // Get the root value from postorder and decrement postorder pointer
        // Get the index of root in inorder from hashmap

        int rootval = postorder[idx];
        int rootidx = map.get(rootval);
        idx--;
        TreeNode node = new TreeNode(rootval);

        // Left recursion where start would be same and end would be index of root in
        // inorder -1
        // Similarly for right recursion
        node.right = helper(postorder, rootidx + 1, end);
        node.left = helper(postorder, start, rootidx - 1);

        return node;
    }
}